### PR TITLE
Feature/metric query result format

### PR DESCRIPTION
**Pull request by abaranau**
metric query result format change according to https://issues.cask.co/browse/CDAP-1243

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    public long getTime() {
+      return time;
+    }
+
+    public long getValue() {
+      return value;
+    }
+  }
+}
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java]: add new line at the end.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * Represents metric query result. This is used for decorating REST API output.
+ */
+// todo: move to cdap-proto along with adding CLI support for metrics querying
+public final class MetricQueryResult {
+  private final long startTime;
+  private final long endTime;
+  private final TimeSeries[] series;
+
+  MetricQueryResult(long startTime, long endTime, TimeSeries[] series) {
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java]: public ?

**Comment by gsps1** [comment]
minor comments. LGTM :+1:

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java (hunk 2 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java]: nah - created within package only

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java (hunk 1 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricQueryResult.java]: done

**Comment by abaranau** [comment]
Please see antoher change: return correct context (that can be copy-pasted into query) when searching metrics and support for "*" (any tag value)

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsHandler.java [hunk 3]
```diff
@@ -121,6 +119,27 @@ public void query(HttpRequest request, HttpResponder responder,
...
+    }
+    String[] tagValues = context.split("\\.");
+
+    // order matters
+    Map<String, String> result = Maps.newLinkedHashMap();
+    for (int i = 0; i < tagValues.length; i += 2) {
+      String tag = tagValues[i];
+      // if odd number, the value for last tag is assumed to be null
+      String val = i + 1 < tagValues.length ? tagValues[i + 1] : null;
+      if ("*".equals(val)) {
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsHandler.java]: use the variable ANY_TAG_VALUE

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java [hunk 4]
```diff
@@ -189,41 +219,56 @@ private String getContext(String nameSpace, String appName, String programType,
...
+                         ".ptp.f.prg.WCounter.run.run1.pr2.splitter",
                        ImmutableList.<String>of());
+
+    // verify "*"
+    verifySearchResult("/v3/metrics/search?target=metric&context=ns.myspace.app.WordCount1" +
+                         ".ptp.f.prg.WordCounter.run.run1.pr2.*",
+                       ImmutableList.<String>of("system.aa", "system.ab", "system.reads",
+                                                "system.writes", "system.zz", "user.reads", "user.writes"));
+    verifySearchResult("/v3/metrics/search?target=metric&context=ns.myspace.app.WordCount1" +
+                         ".ptp.f.prg.*.run.run1",
```
Comment (gsps1) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java]: nit : once we add ANY (*) the next tags doesn't matter. run.run1?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsHandler.java (hunk 3 above)
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsHandler.java]: yes

File: cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java (hunk 4 above)
Comment (abaranau) [inline cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java]: it does matter - it affects filtering ;)

**Comment by gsps1** [comment]
LGTM

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT787-4

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsHandler.java [hunk 5]
```diff
@@ -139,40 +168,70 @@ private void searchChildContextAndRespond(HttpResponder responder, String contex
...
-      //todo : adding null for runId,should we support searching with runId ?
-      contextTags.add(4, new TagValue("run", null));
-    }
     return contextTags;
   }
 
   private Collection<String> searchChildContext(String contextPrefix) throws Exception {
-    CubeExploreQuery searchQuery = new CubeExploreQuery(0, Integer.MAX_VALUE - 1, 1, -1, getContext(contextPrefix));
+    List<TagValue> tagValues = parseTagValues(contextPrefix);
+    toCanonicalContext(tagValues);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/query/MetricsHandler.java]: this is duplicated below

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT787-5 GREEN

**Pull request merged**